enhancements/update/configurable-clusterversion-signature-stores.md

```diff
@@ -0,0 +1,194 @@
+
+
+### User Stories
+
+* As a cluster administrator, I want the cluster's CVO to add new signature store URIs, so that I can have local release signature 
+  verification in air gapped environments.
+
+### Goals
+
+* To verify the release image signature in disconnected environments using OSUS signatures URI.
+* To replace existing signature stores with custom user-provided stores.
+
+### Non-Goals
+
+* This enhancement does not cover how OSUS running in the disconnected environment gets the signatures and if the 
+signatures are validated during the mirroring process.
+
+## Proposal
+
+### Workflow Description
+
+1. Cluster admin will have to add the verification URI to CVO
+2. CVO will reach the newly added URI to verify the release images.
+
+### API Extensions
+
+* CVO will grow a new field, an array, which will be populated with the image verification URI.
+* This will be an array as it'll allow the admin to use multiple locally configured stores. 
+* Existing default signature stores will be used as fallback in-case custom stores don't contain signatures or are offline. 
+* If this property is empty, default stores will be used, similar to how we handle upstream for update service today.
+* Admins will still be able to use the existing config maps for signature verification. 
+* CVO will check the release signatures in the local ConfigMap first and will reach in parallel to other signature stores.
+
+This enhancement proposes to modify the [api/config/v1/types_cluster_version.go](https://github.com/openshift/api/blob/master/config/v1/types_cluster_version.go) as following,
+
+A new field `SignatureStores` will be added to the [`ClusterVersionSpec`](https://github.com/openshift/api/blob/master/config/v1/types_cluster_version.go#L40-L96) structure:
+
+```go
+type ClusterVersionSpec struct {
+	...
+	// SignatureStores contains the upstream URIs to verify release signatures.
+	// By default, CVO will use existing signature stores if this property is empty.
+	// The CVO will check the release signatures in the local ConfigMaps first. It will search for a valid signature 
+	// in these stores in parallel only when local ConfigMaps did not include a valid signature.
+	// Validation will fail if none of the signature stores reply with valid signature before timeout.
+	//
+	// Items in this list should be a valid absolute http/https URI of an upstream signature store as per rfc1738.
+	// +kubebuilder:validation:Format=XValidation:rule="self.all(x, isURL(x))",message="signatureStores must contain only valid absolute URLs per the Go net/url standard"
+	// +listType=set
+	// +optional
+	SignatureStores [32]string `json:"signatureStores,omitempty"`
+}
+```
+
+
+### Risks and Mitigations
+
+* Users will have to make sure that they've cloned all the signatures needed for image verification or use some of the 
+existing stores along with custom stores as missing signatures will cause the verification to fail.
```

Comment:
What happens when the verification fails and how is this exposed to the end user?

Reply:
With a Cluster Bot `launch 4.15.0-ec.1 gcp` cluster, requesting an update to an unsigned image (I'm using a pull-spec that doesn't exist, but because [we verify the signature before attempting to pull the image][1], it doesn't matter whether the image exists or not).

```console
$ oc adm upgrade
Cluster version is 4.15.0-ec.1
...
$ oc adm upgrade --allow-explicit-upgrade --to-image quay.io/openshift-release-dev/ocp-release@sha256:0000000000000000000000000000000000000000000000000000000000000000
warning: The requested upgrade image is not one of the available updates. You have used --allow-explicit-upgrade for the update to proceed anyway
Requested update to release image quay.io/openshift-release-dev/ocp-release@sha256:0000000000000000000000000000000000000000000000000000000000000000
$ oc adm upgrade
Cluster version is 4.15.0-ec.1

ReleaseAccepted=False

  Reason: RetrievePayload
  Message: Retrieving payload failed version="" image="quay.io/openshift-release-dev/ocp-release@sha256:0000000000000000000000000000000000000000000000000000000000000000" failure=The update cannot be verified: unable to verify sha256:0000000000000000000000000000000000000000000000000000000000000000 against keyrings: verifier-public-key-redhat

warning: Cannot display available updates:
  Reason: NoChannel
  Message: The update channel has not been configured.
```

The console squeezes that into a pop-up:

![image](https://github.com/openshift/enhancements/assets/209920/8ae0a5bc-25db-43bb-8fe2-bed2599a9ce1)

There is currently no discussion of the routes attempted to retrieve the signature in ClusterVersion `conditions`, although we do explain the sources when the container starts:

```console
$ oc -n openshift-cluster-version logs -l k8s-app=cluster-version-operator --tail -1 | grep mirror.openshift.com
I1110 22:00:29.431251       1 cvo.go:301] Verifying release authenticity: All release image digests must have GPG signatures from verifier-public-key-redhat (567E347AD0044ADE55BA8A5F199E2F91FD431D51: Red Hat, Inc. (release key 2) <security@redhat.com>, B08B659EE86AF623BC90E8DB938A80CAF21541EB: Red Hat, Inc. (beta key 2) <security@redhat.com>) - will check for signatures in containers/image format at serial signature store wrapping config maps in openshift-config-managed with label "release.openshift.io/verification-signatures", parallel signature store wrapping containers/image signature store under https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release, containers/image signature store under https://storage.googleapis.com/openshift-release/official/signatures/openshift/release
```

We do have _access_ to the details, and [we include them in `acceptedRisks` for folks who force past the verification failure][2].  But they're pretty chatty, and [folks don't always find them accessible][3].  Wrapping the one in Yang's ticket with newlines and indenting to clarify the inverted control tree:

```
[2022-10-24T02:22:58Z: prefix sha256-03a3d3288070259e8cd3e60b3ca7da1c8d410fb3620c44f2e1422dd4fb8f4366 in config map signatures-managed: no more signatures to check,
     2022-10-24T02:22:58Z: unable to retrieve signature from https://storage.googleapis.com/openshift-release/official/signatures/openshift/release/sha256=03a3d3288070259e8cd3e60b3ca7da1c8d410fb3620c44f2e1422dd4fb8f4366/signature-1: no more signatures to check,
     2022-10-24T02:22:58Z: unable to retrieve signature from https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/sha256=03a3d3288070259e8cd3e60b3ca7da1c8d410fb3620c44f2e1422dd4fb8f4366/signature-1: no more signatures to check,
   2022-10-24T02:22:58Z: parallel signature store wrapping containers/image signature store under https://storage.googleapis.com/openshift-release/official/signatures/openshift/release, containers/image signature store under https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release: no more signatures to check,
 2022-10-24T02:22:58Z: serial signature store wrapping config maps in openshift-config-managed with label "release.openshift.io/verification-signatures", parallel signature store wrapping containers/image signature store under https://storage.googleapis.com/openshift-release/official/signatures/openshift/release, containers/image signature store under https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release: no more signatures to check]
```

So that's:

1. There's a serial store that's out of signatures, because its constituent...
    1. ...config map store is out of signatures and its...
    2. ...parallel store is out of signatures.  Because its constituent:
        * storage.googleapis.com is out of signatures and its...
        * mirror.openshift.com is out of signatures.
 
I dunno, it's hard for me to clearly explain this one case in a one-line string, let alone write generic tooling that will always generate clearly-explained strings for a signature lookup that can contain many attempts which could fail in various exciting ways.

[1]: https://github.com/openshift/cluster-version-operator/blob/52b7f32cc1703bd56b6ef6ea7b465870d8af66ad/pkg/cvo/updatepayload.go#L127-L149
[2]: https://github.com/openshift/cluster-version-operator/blob/52b7f32cc1703bd56b6ef6ea7b465870d8af66ad/pkg/cvo/sync_worker.go#L306-L319
[3]: https://issues.redhat.com/browse/OCPBUGS-2819

Reply:
Ok, so, this is complicated, but the required information is exposed, via a condition or via the logs. Given we already have this exposed, I don't think we need to re-invent the wheel here, but might be worth an addendum to look at the UX of this later